### PR TITLE
[GPU] Add pattern to fuse tensor.extract_slice into forall producer 

### PR DESCRIPTION
**Pull request by Max191**
This PR adds a pattern to fuse a consumer tensor.extract_slice into a producer scf.forall op. The transform is added to FuseAndHoistParallelLoops, where it helps to fuse tensor.unpack ops with extract_slice semantics into producer loops. This is needed when targeting MFMA intrinsics for unaligned shapes, and also in generating code for unset encoding ops on GPU. This is a follow up to https://github.com/iree-org/iree/pull/19295, which has the complementing pattern for collapse_shape.

The PR also adds a transform op to keep the long lit tests separate from the FuseAndHoistParallelLoop tests.

### CODE REVIEW

**Comment by Max191** [comment]
This is based on https://github.com/iree-org/iree/pull/19295. Please only review the last commit.

Edit: rebased now.

**Comment by Max191** [comment]
@qedawkins gentle ping. I rebased this PR now that https://github.com/iree-org/iree/pull/19295 has been merged, and it is ready for review.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TransformExtensions/IREEGPUExtensionsOps.td [hunk 1]
```diff
@@ -262,4 +262,38 @@ def FuseCollapseShapeIntoForallOp : Op<Transform_Dialect, "iree.fuse_collapse_sh
...
+     DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+     DeclareOpInterfaceMethods<TransformOpInterface>,
+     ReportTrackingListenerFailuresOpTrait]> {
+  let description = [{
+    Fuses a consumer tensor.extract_slice op into a producer scf.forall op.
+    The users of the block argument for the corresponding forall output operand
+    should be only a tensor.parallel_insert_slice op, and tensor.extract_slice
+    ops that extract an equivalent subset. After the fusion, the output of the
+    forall will be an equal subset slice of the original output, and all users
+    of this block arg will be clamped to the slice size. Additional tensor.pad
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TransformExtensions/IREEGPUExtensionsOps.td]: Docs look a little out of date, you aren't generating pads anymore.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TransformExtensions/test/transform_fuse_extract_slice_with_forall.mlir [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+// CHECK-LABEL: func @fuse_extract_slice_into_multi_result_forall
+
+//       CHECK:   %[[FORALL_RESULT:.+]]:2 = scf.forall {{.*}} -> (tensor<?xf32>, tensor<8xf32>) {
+//       CHECK:     scf.forall.in_parallel {
+//   CHECK-DAG:       tensor.parallel_insert_slice {{.*}} : tensor<?xf32> into tensor<?xf32>
+//   CHECK-DAG:       tensor.parallel_insert_slice {{.*}} : tensor<2xf32> into tensor<8xf32>
+//       CHECK:     }
+//       CHECK:   } {mapping = [#gpu.thread<x>]}
+//       CHECK:   return %[[FORALL_RESULT]]#0, %[[FORALL_RESULT]]#1
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TransformExtensions/test/transform_fuse_extract_slice_with_forall.mlir:279]: Can you add a test with non-zero offset? (thanks for the nice tests)

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp [hunk 3]
```diff
@@ -518,6 +516,263 @@ fuseCollapseShapeIntoProducerForall(RewriterBase &rewriter,
...
+         llvm::zip_equal(offsets, sizes, upperBoundSizes)) {
+      AffineExpr d0, d1, d2;
+      auto ctx = rewriter.getContext();
+      bindDims(ctx, d0, d1, d2);
+      auto ubClampMap = AffineMap::get(3, 0, {d0 - d1, d2}, ctx);
+      auto lbClampMap = rewriter.getMultiDimIdentityMap(2);
+      auto ubClamped = affine::makeComposedFoldedAffineMin(
+          rewriter, loc, ubClampMap, {ub, offset, size});
+      auto lbClamped = affine::makeComposedFoldedAffineMax(
+          rewriter, loc, lbClampMap, {ubClamped, rewriter.getIndexAttr(0)});
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp]: nit: spell out types for min/max

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp [hunk 4]
```diff
@@ -518,6 +516,263 @@ fuseCollapseShapeIntoProducerForall(RewriterBase &rewriter,
...
+static FailureOr<tensor::ParallelInsertSliceOp>
+clampParallelInsertSliceOp(RewriterBase &rewriter,
+                           tensor::ParallelInsertSliceOp parallelInsertOp,
+                           SmallVector<OpFoldResult> upperBoundSizes) {
+  // Clamp the sizes of the parallel insert source.
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(parallelInsertOp.getParallelCombiningParent());
+  Location loc = parallelInsertOp.getParallelCombiningParent()->getLoc();
+
+  auto clampSizes = [&](SmallVector<OpFoldResult> offsets,
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp]: Do we need the lambda here? It's only used once.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp [hunk 5]
```diff
@@ -518,6 +516,263 @@ fuseCollapseShapeIntoProducerForall(RewriterBase &rewriter,
...
+                        SmallVector<OpFoldResult> sizes, Location loc) {
+    SmallVector<OpFoldResult> clampedSizes;
+    for (auto [offset, size, ub] :
+         llvm::zip_equal(offsets, sizes, upperBoundSizes)) {
+      AffineExpr d0, d1, d2;
+      auto ctx = rewriter.getContext();
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp]:
```suggestion
      MLIRContext *ctx = rewriter.getContext();
```

**Review by qedawkins — APPROVED** [review]
nice work! Also don't feel obligated to add transform ops, they do add extra code to maintain.
